File: packages/ts-shared/src/result.ts
export type Result<T, E = Error> = { ok: true; value: T } | { ok: false; error: E };

export function ok<T>(value: T): Result<T, never> {
  return { ok: true, value };
}

export function err<E>(error: E): Result<never, E> {
  return { ok: false, error };
}

/** Wraps a throwing sync function, turning an exception into an `err`. */
export function tryCatch<T>(fn: () => T): Result<T> {
  try {
    return ok(fn());
  } catch (error) {
    return err(error instanceof Error ? error : new Error(String(error)));
  }
}

File: packages/py-shared/src/py_shared/result.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class Ok[T]:
    value: T
    ok: bool = True


@dataclass(frozen=True, slots=True)
class Err[E]:
    error: E
    ok: bool = False


type Result[T, E] = Ok[T] | Err[E]


def try_catch[T](fn: Callable[[], T]) -> Result[T, Exception]:
    """Calls `fn`, turning a raised exception into an `Err` instead of propagating it."""
    try:
        return Ok(fn())
    except Exception as exc:  # intentionally broad, mirrors ts-shared's tryCatch
        return Err(exc)
